README.md
# neomorphicdesign

A set of neomorphic design based widgets.

# How to use ?

## Import it in your pubspec.yaml
```yaml
neomorphic_design:
  git:
    url: git@github.com:jenjiskan/neomorphic-design-package.git
```

## Initialize Neomorphic theme in your main function
```dart
void main() {
  NeomorphicTheme.init(
      baseColor: Color(0xFF243441),
      darkColor: Colors.black.withAlpha(100),
      lightColor: Colors.white.withAlpha(25),
      shadowBlurRadius: 16,
      shadowOffset: 5,
      textStyleDark: TextStyle(color: Colors.black.withAlpha(100)),
      textStyleLight: TextStyle(color: Colors.white.withAlpha(25)),
      textStyleButtonLabel: TextStyle(color: Colors.white.withAlpha(25), fontWeight: FontWeight.bold),
      textStyleTextField: TextStyle(color: Colors.white.withAlpha(200))
  );
  runApp(MyApp());
}
```
## Use widgets as following example
```dart
class Home extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: SafeArea(
        child: ListView(
          children: <Widget>[
            Padding(padding: EdgeInsets.all(20),child: Text('Buttons variants', style: Theme.of(context).textTheme.title.copyWith(color: Colors.white.withAlpha(25)),)),
            NeomorphicRaisedButton(
                label: 'BUTTON',
                horizontalMargin: 20.0, verticalMargin: 20.0,
                horizontalPadding: 16.0, verticalPadding: 16.0,
                leadingIcon: Icon(
                  Icons.four_k,
                  color: Colors.white.withAlpha(25),
                )
            ),
            NeomorphicRaisedButton(
                label: 'BUTTON',
                horizontalMargin: 20.0, verticalMargin: 20.0,
                horizontalPadding: 16.0, verticalPadding: 16.0,
                trailingIcon: Icon(
                  Icons.four_k,
                  color: Colors.white.withAlpha(25),
                )
            ),
            NeomorphicRaisedButton(
                label: 'BUTTON',
                horizontalMargin: 20.0, verticalMargin: 20.0,
                horizontalPadding: 16.0, verticalPadding: 16.0,
                leadingIcon: Icon(
                  Icons.four_k,
                  color: Colors.white.withAlpha(25),
                ),
                trailingIcon: Icon(
                  Icons.four_k,
                  color: Colors.white.withAlpha(25),
                )
            ),
            NeomorphicRaisedButton(
                label: 'BUTTON',
                horizontalMargin: 20.0, verticalMargin: 20.0,
                horizontalPadding: 16.0, verticalPadding: 16.0
            ),
            Padding(padding: EdgeInsets.all(20),child: Text('Checkbox variants', style: Theme.of(context).textTheme.title.copyWith(color: Colors.white.withAlpha(25)),)),SizedBox(
              width: double.infinity,
              height: 20,
            ),
            Row(
              mainAxisAlignment: MainAxisAlignment.spaceAround,
              children: <Widget>[
                NeomorphicCheckbox(
                    isChecked: false,
                    uncheckedColor: Colors.red,
                    checkedColor: Colors.lightGreen,
                    height: 32,
                    width: 32,
                    borderRadius: 16.0
                ),
                NeomorphicCheckbox(
                  isChecked: false,
                  height: 32,
                  width: 32,
                  checkedIcon: Icon(
                    Icons.check,
                    color: Colors.lightGreen,
                    size: 16,
                  ),
                  uncheckedIcon: Icon(
                    Icons.close,
                    color: Colors.red.withAlpha(200),
                    size: 16,
                  ),
                  borderRadius: 16.0,
                  onCheck: (c) => print('isCheched: $c'),
                ),
                NeomorphicCheckbox(
                  isChecked: false,
                  height: 32,
                  width: 32,
                  checkedIcon: Icon(
                    Icons.check,
                    color: getIt<NeomorphicTheme>().lightColor,
                    size: 16,
                  ),
                  uncheckedIcon: Icon(
                    Icons.close,
                    color: getIt<NeomorphicTheme>().lightColor,
                    size: 16,
                  ),
                  borderRadius: 16.0,
                  onCheck: (c) => print('isCheched: $c'),
                ),
                NeomorphicCheckbox(
                  isChecked: false,
                  height: 32,
                  width: 32,
                  checkedIcon: Icon(
                    Icons.check,
                    color: Colors.white.withAlpha(25),
                    size: 16,
                  ),
                  uncheckedIcon: Icon(
                    Icons.close,
                    color: Colors.white.withAlpha(25),
                    size: 16,
                  ),
                  borderRadius: 16.0,
                  uncheckedColor: Colors.red,
                  checkedColor: Colors.lightGreen,
                  onCheck: (c) => print('isCheched: $c'),
                )
              ],
            ),
            SizedBox(
              width: double.infinity,
              height: 50,
            ),
            Row(
              mainAxisAlignment: MainAxisAlignment.spaceAround,
              children: <Widget>[
                NeomorphicCheckbox(
                  isChecked: true,
                  uncheckedColor: Colors.red,
                  checkedColor: Colors.lightGreen,
                  height: 32,
                  width: 32,
                  borderRadius: 8.0,
                ),
                NeomorphicCheckbox(
                  isChecked: true,
                  height: 32,
                  width: 32,
                  checkedIcon: Icon(
                    Icons.check,
                    color: Colors.lightGreen,
                    size: 16,
                  ),
                  uncheckedIcon: Icon(
                    Icons.close,
                    color: Colors.red.withAlpha(200),
                    size: 16,
                  ),
                  borderRadius: 8.0,
                  onCheck: (c) => print('isCheched: $c'),
                ),
                NeomorphicCheckbox(
                  isChecked: true,
                  height: 32,
                  width: 32,
                  checkedIcon: Icon(
                    Icons.check,
                    color: Colors.white.withAlpha(25),
                    size: 16,
                  ),
                  uncheckedIcon: Icon(
                    Icons.close,
                    color: Colors.white.withAlpha(25),
                    size: 16,
                  ),
                  borderRadius: 8.0,
                  onCheck: (c) => print('isCheched: $c'),
                ),
                NeomorphicCheckbox(
                  isChecked: true,
                  height: 32,
                  width: 32,
                  checkedIcon: Icon(
                    Icons.check,
                    color: Colors.white.withAlpha(25),
                    size: 16,
                  ),
                  uncheckedIcon: Icon(
                    Icons.close,
                    color: Colors.white.withAlpha(25),
                    size: 16,
                  ),
                  borderRadius: 8.0,
                  uncheckedColor: Colors.red,
                  checkedColor: Colors.lightGreen,
                  onCheck: (c) => print('isCheched: $c'),
                )
              ],
            ),
            SizedBox(
              width: double.infinity,
              height: 50,
            ),
            Padding(padding: EdgeInsets.all(20),child: Text('TextFields variants', style: Theme.of(context).textTheme.title.copyWith(color: Colors.white.withAlpha(25)),)),
            NeomorphicTextFormField(
              horizontalMargin: 20.0, verticalMargin: 20.0,
              borderRadius: 50,
              horizontalPadding: 25,
              hintText: 'hint',
              autovalidate: true,
              validator: (v)=> v.length < 5 ? 'error message here' : null,
            ),
            NeomorphicTextFormField(
              horizontalMargin: 20.0, verticalMargin: 20.0,
              borderRadius: 50,
              horizontalPadding: 25,
              hintText: 'hint',
            ),
            NeomorphicTextFormField(
              horizontalMargin: 20.0, verticalMargin: 20.0,
              borderRadius: 50,
              horizontalPadding: 25,
              hintText: 'hint',
              initialValue: 'text inserted here',
            ),
            NeomorphicTextFormField(
              horizontalMargin: 20.0, verticalMargin: 20.0,
              borderRadius: 50,
              horizontalPadding: 25,
              hintText: 'hint',
              maxLines: 8,
              maxLength: 500,
            ),
            SizedBox(
              width: double.infinity,
              height: 32,
            ),
            Padding(padding: EdgeInsets.all(20),child: Text('Container variants', style: Theme.of(context).textTheme.title.copyWith(color: Colors.white.withAlpha(25)),)),
            NeomorphicContainer(
              horizontalMargin: 20.0,
              verticalMargin: 20.0,
              containerType: ContainerType.positive,
            ),
            NeomorphicContainer(
              horizontalMargin: 20.0,
              verticalMargin: 20.0,
              containerType: ContainerType.negative,
            )
          ],
        ),
      ),
    );
  }
}
```

CHANGELOG.md
## [0.0.1] - 22/03//2020.

* First release.

## [0.0.2] - 23/03/2020.

* Internalization of theme.


lib/neomorphic_theme.dart
import 'package:flutter/material.dart';
import 'package:get_it/get_it.dart';

GetIt getIt = GetIt.instance;

class NeomorphicTheme {
  final Color lightColor;
  final Color darkColor;
  final Color baseColor;
  final TextStyle textStyleLight;
  final TextStyle textStyleDark;
  final TextStyle textStyleButtonLabel;
  final TextStyle textStyleTextField;
  final double shadowOffset;
  final double shadowBlurRadius;

  NeomorphicTheme._(
    this.lightColor,
    this.darkColor,
    this.baseColor,
    this.textStyleLight,
    this.textStyleDark,
    this.textStyleButtonLabel,
    this.textStyleTextField,
    this.shadowOffset,
    this.shadowBlurRadius,
  );

  static void init({
    Color lightColor,
    Color darkColor,
    Color baseColor,
    TextStyle textStyleLight,
    TextStyle textStyleDark,
    TextStyle textStyleButtonLabel,
    TextStyle textStyleTextField,
    double shadowOffset = 5.0,
    double shadowBlurRadius = 16.0
  }) => getIt.registerSingleton<NeomorphicTheme>(NeomorphicTheme._(
      lightColor,
      darkColor,
      baseColor,
      textStyleLight,
      textStyleDark,
      textStyleButtonLabel,
      textStyleTextField,
      shadowOffset,
      shadowBlurRadius,
  ), signalsReady: true);
}

lib/neomorphic_check_box.dart
import 'dart:math';

import 'package:flutter/material.dart';
import 'package:neomorphic_design/neomorphic_theme.dart';

class NeomorphicCheckbox extends StatefulWidget {
  static const _defaultCheckboxDimension = 24.0;

  final bool isChecked;
  final Function onCheck;
  final double borderRadius;
  final double height;
  final double width;
  final double horizontalMargin;
  final double verticalMargin;
  final double horizontalPadding;
  final double verticalPadding;
  final Icon checkedIcon;
  final Icon uncheckedIcon;
  final Color checkedColor;
  final Color uncheckedColor;
  final Duration checkAnimationDuration;
  final NeomorphicTheme theme = getIt<NeomorphicTheme>();

  NeomorphicCheckbox(
      {Key key,
        this.isChecked,
        this.borderRadius = 8.0,
        this.height = _defaultCheckboxDimension,
        this.width = _defaultCheckboxDimension,
        this.horizontalMargin = .0,
        this.verticalMargin= .0,
        this.horizontalPadding = 8.0,
        this.verticalPadding = 8.0,
        this.onCheck,
        this.checkedIcon,
        this.uncheckedIcon,
        this.checkedColor,
        this.uncheckedColor,
        this.checkAnimationDuration = const Duration(milliseconds: 60)}) : super(key: key) {
    if (checkedIcon != null && checkedIcon.size != min(width, height) / 2 ||
        uncheckedIcon != null && uncheckedIcon.size != min(width, height) / 2) {
      throw Exception('Icon size has to be half of the minor between height and width');
    }
  }

  @override
  _NeomorphicCheckboxState createState() => _NeomorphicCheckboxState();
}

class _NeomorphicCheckboxState extends State<NeomorphicCheckbox> {
  bool _isChecked;
  List<BoxShadow> _shadow;

  @override
  void initState() {
    _isChecked = widget.isChecked;
    _shadow = _isChecked ? _checked() : _unchecked();
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return GestureDetector(
        onTapDown: onTapDown,
        onTapUp: onTapUp,
        onTapCancel: onTapCancel,
        child: AnimatedContainer(
            alignment: Alignment.center,
            margin: EdgeInsets.symmetric(horizontal: widget.horizontalMargin, vertical: widget.verticalMargin),
            padding: EdgeInsets.symmetric(horizontal: widget.horizontalPadding, vertical: widget.verticalPadding),
            duration: widget.checkAnimationDuration,
            height: widget.height,
            width: widget.width,
            decoration: BoxDecoration(borderRadius: BorderRadius.circular(widget.borderRadius), boxShadow: _shadow),
            child: Center(
              child: _isChecked ? widget.checkedIcon : widget.uncheckedIcon,
            )));
  }

  void onTapDown(TapDownDetails detail) {
    setState(() {
      _shadow = _isChecked ? _unchecked() : _checked();
      _isChecked = !_isChecked;
    });
  }

  void onTapUp(TapUpDetails detail) {
    widget.onCheck?.call(_isChecked);
  }

  void onTapCancel() {
    setState(() => _shadow = _isChecked ? _checked() : _unchecked());
  }

  List<BoxShadow> _unchecked() => [
    BoxShadow(
        color: widget.theme.darkColor,
        offset: Offset(widget.theme.shadowOffset, widget.theme.shadowOffset),
        blurRadius: widget.theme.shadowBlurRadius),
    BoxShadow(
        color: widget.theme.lightColor,
        offset: Offset(-widget.theme.shadowOffset, -widget.theme.shadowOffset),
        blurRadius: widget.theme.shadowBlurRadius),
    BoxShadow(color: widget.uncheckedColor ?? widget.theme.baseColor),
    BoxShadow(color: widget.theme.baseColor, spreadRadius: -1)
  ];

  List<BoxShadow> _checked() => [
    BoxShadow(
        color: widget.theme.lightColor,
        offset: Offset(widget.theme.shadowOffset, widget.theme.shadowOffset),
        blurRadius: widget.theme.shadowBlurRadius),
    BoxShadow(
        color: widget.theme.darkColor,
        offset: Offset(-widget.theme.shadowOffset, -widget.theme.shadowOffset),
        blurRadius: widget.theme.shadowBlurRadius),
    BoxShadow(color: widget.checkedColor ?? widget.theme.baseColor),
    BoxShadow(color: widget.theme.baseColor, spreadRadius: -1)
  ];
}


lib/neomorphic_container.dart
import 'package:flutter/material.dart';
import 'package:neomorphic_design/neomorphic_theme.dart';

class NeomorphicContainer extends StatelessWidget {

  final double borderRadius;
  final double height;
  final double width;
  final double horizontalMargin;
  final double verticalMargin;
  final double horizontalPadding;
  final double verticalPadding;
  final Alignment alignment;
  final ContainerType containerType;
  final NeomorphicTheme theme = getIt<NeomorphicTheme>();

  NeomorphicContainer({
    Key key,
    @required this.containerType,
    this.borderRadius = 8.0,
    this.height,
    this.width,
    this.alignment = Alignment.topLeft, 
    this.horizontalMargin = .0, 
    this.verticalMargin = .0, 
    this.horizontalPadding = .0, 
    this.verticalPadding = .0
  }): super(key: key);


  @override
  Widget build(BuildContext context) => containerType == ContainerType.positive
      ? _positive(context)
      : _negative(context);

  Widget _negative(BuildContext context) {
    return Padding(
      padding: EdgeInsets.symmetric(horizontal: horizontalMargin, vertical: verticalMargin),
      child: ClipRRect(
        clipBehavior: Clip.antiAlias,
        borderRadius: BorderRadius.circular(borderRadius),
        child: Container(
          padding: EdgeInsets.symmetric(horizontal: horizontalPadding, vertical: verticalMargin),
          height: height,
          width: width,
          decoration: BoxDecoration(borderRadius: BorderRadius.circular(borderRadius), boxShadow: [
            BoxShadow(color: Theme.of(context).primaryColorLight, offset: Offset(theme.shadowOffset, theme.shadowOffset), blurRadius: theme.shadowBlurRadius),
            BoxShadow(color: Theme.of(context).primaryColorDark, offset: Offset(-theme.shadowOffset, -theme.shadowOffset), blurRadius: theme.shadowBlurRadius),
            BoxShadow(color: Theme.of(context).primaryColor, blurRadius: theme.shadowBlurRadius),
          ]),
        ),
      ),
    );
  }

  Widget _positive(BuildContext context) {
    return Container(
      margin: EdgeInsets.symmetric(horizontal: horizontalMargin, vertical: verticalMargin),
      padding: EdgeInsets.symmetric(horizontal: horizontalPadding, vertical: verticalMargin),
      height: height,
      width: width,
      decoration: BoxDecoration(color: Theme.of(context).primaryColor, borderRadius: BorderRadius.circular(borderRadius), boxShadow: [
        BoxShadow(color: Theme.of(context).primaryColorDark, offset: Offset(theme.shadowOffset, theme.shadowOffset), blurRadius: theme.shadowBlurRadius),
        BoxShadow(color: Theme.of(context).primaryColorLight, offset: Offset(-theme.shadowOffset, -theme.shadowOffset), blurRadius: theme.shadowBlurRadius)
      ]),
    );
  }

}

enum ContainerType {
  positive,
  negative
}


lib/neomorphic_raised_button.dart
import 'package:flutter/material.dart';
import 'package:neomorphic_design/neomorphic_theme.dart';

class NeomorphicRaisedButton extends StatefulWidget {

  final String label;
  final Function onPressed;
  final double borderRadius;
  final double height;
  final double width;
  final double horizontalMargin;
  final double verticalMargin;
  final double horizontalPadding;
  final double verticalPadding;
  final TextAlign textAlignment;
  final Icon leadingIcon;
  final Icon trailingIcon;
  final Duration pressAnimationDuration;
  final NeomorphicTheme theme = getIt<NeomorphicTheme>();

  NeomorphicRaisedButton({
    Key key,
    this.label,
    this.onPressed,
    this.borderRadius = 8.0,
    this.height,
    this.width,
    this.horizontalMargin = .0,
    this.verticalMargin= .0,
    this.horizontalPadding = 8.0,
    this.verticalPadding = 8.0,
    this.leadingIcon,
    this.trailingIcon,
    this.textAlignment = TextAlign.center,
    this.pressAnimationDuration = const Duration(milliseconds: 60)
  }): super(key: key);

  double get _pressedShadowOffset => (theme.shadowOffset * 2) / 5;

  @override
  _NeomorphicRaisedButtonState createState() => _NeomorphicRaisedButtonState();
}

class _NeomorphicRaisedButtonState extends State<NeomorphicRaisedButton> {

  List<BoxShadow> _shadow;

  @override
  void initState() {
    _shadow = _normal();
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return GestureDetector(
      onTapDown: onTapDown,
      onTapUp: onTapUp,
      onTapCancel: onTapCancel,
      child: AnimatedContainer(
          alignment: Alignment.center,
          margin: EdgeInsets.symmetric(horizontal: widget.horizontalMargin, vertical: widget.verticalMargin),
          padding: EdgeInsets.symmetric(horizontal: widget.horizontalPadding, vertical: widget.verticalPadding),
          duration: widget.pressAnimationDuration,
          height: widget.height,
          width: widget.width,
          decoration: BoxDecoration(
              color: widget.theme.baseColor,
              borderRadius: BorderRadius.circular(widget.borderRadius),
              boxShadow: _shadow
          ),
          child: Row(
            children: <Widget>[
              widget.leadingIcon,
              Expanded(
                  child: Text(widget.label, textAlign: widget.textAlignment, style: widget.theme.textStyleButtonLabel,)
              ),
              widget.trailingIcon
            ]..removeWhere((item)=> item == null),
          )
      ),
    );
  }

  void onTapDown(TapDownDetails detail) {
    setState(() => _shadow = _pressed());
  }

  void onTapUp(TapUpDetails detail) {
    setState(() => _shadow = _normal());
    widget.onPressed?.call();
  }

  void onTapCancel() {
    setState(() => _shadow = _normal());
  }

  List<BoxShadow> _normal() => [
    BoxShadow(
        color: widget.theme.darkColor,
        offset: Offset(widget.theme.shadowOffset, widget.theme.shadowOffset),
        blurRadius: widget.theme.shadowBlurRadius
    ),
    BoxShadow(
        color: widget.theme.lightColor,
        offset: Offset(-widget.theme.shadowOffset, -widget.theme.shadowOffset),
        blurRadius: widget.theme.shadowBlurRadius
    )
  ];

  List<BoxShadow> _pressed() => [
    BoxShadow(
        color: widget.theme.darkColor,
        offset: Offset(widget._pressedShadowOffset, widget._pressedShadowOffset),
        blurRadius: widget.theme.shadowBlurRadius
    ),
    BoxShadow(
        color: widget.theme.lightColor,
        offset: Offset(-widget._pressedShadowOffset, -widget._pressedShadowOffset),
        blurRadius: widget.theme.shadowBlurRadius
    )
  ];

}


lib/neomorphic_text_field.dart
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:neomorphic_design/neomorphic_theme.dart';

class NeomorphicTextFormField extends StatefulWidget {
  final TextEditingController controller;
  final String initialValue;
  final FocusNode focusNode;
  final TextInputType keyboardType;
  final TextCapitalization textCapitalization;
  final TextInputAction textInputAction;
  final StrutStyle strutStyle;
  final TextDirection textDirection;
  final TextAlign textAlign;
  final TextAlignVertical textAlignVertical;
  final bool autofocus;
  final bool readOnly;
  final ToolbarOptions toolbarOptions;
  final bool showCursor;
  final bool obscureText;
  final bool autocorrect;
  final bool enableSuggestions;
  final bool autovalidate;
  final bool maxLengthEnforced;
  final int maxLines;
  final int minLines;
  final bool expands;
  final int maxLength;
  final ValueChanged<String> onChanged;
  final GestureTapCallback onTap;
  final VoidCallback onEditingComplete;
  final ValueChanged<String> onFieldSubmitted;
  final FormFieldSetter<String> onSaved;
  final FormFieldValidator<String> validator;
  final List<TextInputFormatter> inputFormatters;
  final bool enabled;
  final double cursorWidth;
  final Radius cursorRadius;
  final Color cursorColor;
  final Brightness keyboardAppearance;
  final EdgeInsets scrollPadding;
  final bool enableInteractiveSelection;
  final InputCounterWidgetBuilder buildCounter;
  final String hintText;

  //neomorphic properties
  final _animationDuration = Duration(milliseconds: 60);
  final double borderRadius;
  final double width;
  final double horizontalMargin;
  final double verticalMargin;
  final double verticalPadding;
  final double horizontalPadding;
  final Icon leadingIcon;
  final Icon trailingIcon;
  final NeomorphicTheme theme = getIt<NeomorphicTheme>();

  NeomorphicTextFormField({
    Key key,
    this.hintText,
    this.controller,
    this.initialValue,
    FocusNode focusNode,
    this.keyboardType,
    this.textCapitalization = TextCapitalization.none,
    this.textInputAction,
    this.strutStyle,
    this.textDirection,
    this.textAlign = TextAlign.start,
    this.textAlignVertical,
    this.autofocus = false,
    this.readOnly = false,
    this.toolbarOptions,
    this.showCursor,
    this.obscureText = false,
    this.autocorrect = true,
    this.enableSuggestions = true,
    this.autovalidate = false,
    this.maxLengthEnforced = true,
    this.maxLines = 1,
    this.minLines,
    this.expands = false,
    this.maxLength,
    this.onChanged,
    this.onTap,
    this.onEditingComplete,
    this.onFieldSubmitted,
    this.onSaved,
    this.validator,
    this.inputFormatters,
    this.enabled = true,
    this.cursorWidth = 2.0,
    this.cursorRadius,
    this.cursorColor,
    this.keyboardAppearance,
    this.scrollPadding = const EdgeInsets.all(20.0),
    this.enableInteractiveSelection = true,
    this.buildCounter,
    this.leadingIcon,
    this.trailingIcon,
    this.borderRadius = 8.0,
    this.width,
    this.horizontalMargin = .0,
    this.verticalMargin= .0,
    this.verticalPadding = 4,
    this.horizontalPadding = 16,
  })  : this.focusNode = focusNode ??= FocusNode(),
        super(key: key);

  @override
  _NeomorphicTextFormFieldState createState() => _NeomorphicTextFormFieldState();
}

class _NeomorphicTextFormFieldState extends State<NeomorphicTextFormField> {
  List<BoxShadow> _shadow;
  TextFormField _textFormField;
  @override
  void initState() {
    _shadow = _normal();
    widget.focusNode.addListener(() {
      setState(() => widget.focusNode.hasFocus ? _shadow = _focused() : _normal());
    });
    _textFormField = TextFormField(
      decoration: InputDecoration(border: InputBorder.none, hintText: widget.hintText, hintStyle: widget.theme.textStyleLight, counterStyle: widget.theme.textStyleLight),
      controller: widget.controller,
      initialValue: widget.initialValue,
      focusNode: widget.focusNode,
      keyboardType: widget.keyboardType,
      textCapitalization: widget.textCapitalization,
      textInputAction: widget.textInputAction,
      style: widget.theme.textStyleTextField,
      strutStyle: widget.strutStyle,
      textDirection: widget.textDirection,
      textAlign: widget.textAlign,
      textAlignVertical: widget.textAlignVertical,
      autofocus: widget.autofocus,
      readOnly: widget.readOnly,
      toolbarOptions: widget.toolbarOptions,
      showCursor: widget.showCursor,
      obscureText: widget.obscureText,
      autocorrect: widget.autocorrect,
      enableSuggestions: widget.enableSuggestions,
      autovalidate: widget.autovalidate,
      maxLengthEnforced: widget.maxLengthEnforced,
      maxLines: widget.maxLines,
      minLines: widget.minLines,
      expands: widget.expands,
      maxLength: widget.maxLength,
      onChanged: widget.onChanged,
      onTap: widget.onTap,
      onEditingComplete: widget.onEditingComplete,
      onFieldSubmitted: widget.onFieldSubmitted,
      onSaved: widget.onSaved,
      validator: widget.validator,
      inputFormatters: widget.inputFormatters,
      enabled: widget.enabled,
      cursorWidth: widget.cursorWidth,
      cursorRadius: widget.cursorRadius,
      cursorColor: widget.cursorColor,
      keyboardAppearance: widget.keyboardAppearance,
      scrollPadding: widget.scrollPadding,
      enableInteractiveSelection: widget.enableInteractiveSelection,
      buildCounter: widget.buildCounter,
    );
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return Padding(
      padding: EdgeInsets.symmetric(horizontal: widget.horizontalMargin, vertical: widget.verticalMargin),
      child: ClipRRect(
        clipBehavior: Clip.antiAlias,
        borderRadius: BorderRadius.circular(widget.borderRadius),
        child: AnimatedContainer(
          alignment: Alignment.center,
          padding: EdgeInsets.symmetric(vertical: widget.verticalPadding, horizontal: widget.horizontalPadding),
          duration: widget._animationDuration,
          width: widget.width,
          decoration: BoxDecoration(borderRadius: BorderRadius.circular(widget.borderRadius), boxShadow: _shadow),
          child: _textFormField,
        ),
      ),
    );
  }

  List<BoxShadow> _normal() => [
        BoxShadow(
            color: widget.theme.lightColor,
            offset: Offset(widget.theme.shadowOffset, widget.theme.shadowOffset),
            blurRadius: widget.theme.shadowBlurRadius),
        BoxShadow(
            color: widget.theme.darkColor,
            offset: Offset(-widget.theme.shadowOffset, -widget.theme.shadowOffset),
            blurRadius: widget.theme.shadowBlurRadius),
        BoxShadow(color: widget.theme.baseColor, blurRadius: widget.theme.shadowBlurRadius),
      ];

  List<BoxShadow> _focused() => [
        BoxShadow(
            color: widget.theme.lightColor,
            offset: Offset(widget.theme.shadowOffset, widget.theme.shadowOffset),
            blurRadius: widget.theme.shadowBlurRadius,
            spreadRadius: -2),
        BoxShadow(
            color: widget.theme.darkColor,
            offset: Offset(-widget.theme.shadowOffset, -widget.theme.shadowOffset),
            blurRadius: widget.theme.shadowBlurRadius,
            spreadRadius: -2),
        BoxShadow(color: widget.theme.baseColor, spreadRadius: -3, blurRadius: widget.theme.shadowBlurRadius),
      ];
}
